Prose/ProseLib/Sources/App/AppReducer.swift

```diff
@@ -95,18 +96,19 @@ public let appReducer: Reducer<
 // MARK: State
 
 public struct AppState: Equatable {
-    var route: AppRoute
```

Comment:
There was absolutely no reason to remove `AppRoute`. Quite the opposite.

1. We want to have uniformity in how we handle navigation throughout the app.
2. To do so we want to use the best representation for a particular state. In this case you have introduced the potential for creating two invalid states, either where `main` and `auth` are nil or where they're both non-nil. That was prevented and enforced by the compiler before. [See also](https://en.wikipedia.org/wiki/Algebraic_data_type).

Reply:
Hold on. I might be dead wrong here. Need to think about this for a bit.

Reply:
I agree with you, the thing is "they're both non-nil" is a valid state.

When a user logs into a new account, we keep the current view as is, not as a placeholder, and we show the auth screen onto it.

I could do

```swift
enum AppRoute {
  case auth(main: MainState?, auth: AuthState)
  case main(MainState)
}
```

but I didn't feel it was worth the clutter in the reducer.

Reply:
Ok, that was non-sense. Both routes/states are not mutually exclusive anymore. The main state is always present while  the authentication flow can be presented additionally.

That being said, I believe there's still room for improvement.

I mentioned that the main state is always present. This kind of gives away that `main` shouldn't be optional. Currently you're using it as a marker to let the UI know if it should enable the redaction on the main window **and** if it should create a mock `Store` so that the `mainWindowReducer` and its children cannot do anything. To do so you even had to stub out the environment all the way down. The last part shouldn't usually be this hard, since a child environment is derived from its parent environment. The code currently doesn't work this way, but that's a different problem.

The thing is that you now have logic in your UI and if you'd wanted to unit test that the environment passed to the `mainWindowReducer` actually does nothing that would be very hard.

Therefore my suggestion is to move the logic to the reducer. We can disable the `mainWindowReducer` while still leaving the environment intact by creating a higher-order reducer that does that for us.

But first we need an indicator when to disable it. Easiest solution in the current state of things:

```swift
extension AppState {
  var mainWindowDisabled: Bool {
    self.auth != nil
  }
}
```

Note that this might not be true in the future. I don't know about our desired UX when our user already has an account and logs into another one. Do we have a live main window below the sheet or do we always disable and redact? However that condition will be easy enough to change. 

The `mainWindowDisabled` property would then also be used by the `AppScene` to activate the redaction. Simple logic there and testable by a snapshot test.

The higher-level reducer looks like this:

```swift
extension Reducer where State == AppState, Action == AppAction, Environment == AppEnvironment {
  func disabled(when isDisabled: @escaping (AppState) -> Bool) -> Self {
    Reducer { state, action, environment in
      guard !isDisabled(state) else {
        return .none
      }
      return self(&state, action, environment)
    }
  }
}
```

and would be used like so…

```swift
public let appReducer: Reducer<
    AppState,
    AppAction,
    AppEnvironment
> = Reducer.combine([
    …,
    mainWindowReducer.optional().pullback(
        state: \AppState.main,
        action: /AppAction.main,
        environment: { $0.main }
    ).disabled(when: \.mainWindowDisabled),
    Reducer { state, action, environment in
    …
```

Now you can unit test this by sending actions to the `mainWindowReducer` and assert that its state doesn't change/the environment is not touched.

Don't forget to remove the now unused `AppRoute`.

Reply:
Oh, and by the way I'm well aware of algebraic data types… and I tend to use them too much 🙈

Reply:
I didn't see your last message when I sent mine, sorry for the unrelated answer

Reply:
Ok, now I understand your proposition, and it's a good idea

Reply:
> The last part shouldn't usually be this hard, since a child environment is derived from its parent environment. The code currently doesn't work this way, but that's a different problem.

Yes, I know, but sometimes deriving can be very verbose (that's why SwiftUI introduced `EnvironmentValues`… how I wish we had that for `struct`s 😅)

Reply:
Not deriving destroys dependency injection.